# prenatal_brain_sarscov2
code for bulk rna-seq analysis

related to doi: https://doi.org/10.1101/2021.01.17.427024

_Bulk RNA sequencing and differential expression analysis_
RNA was extracted from a GW21 and GW23 primary cortical sample after 3 days of infection, as before.  RNA was sequenced through Geneviz using 150bp paired end RNA-seq on a Illumina HiSeq  platform. Reads were aligned and quantified using kallisto and collapsed to gene level quantifications with tximport (47, 48). Differential expression analysis was performed using limma/voom, accounting for 2 individuals with blocked replicates within each individual. P values were corrected for multiple comparisons using the Benjamini-Hochberg false-discovery rate adjustment (48, 49).

_Pathway and gene set enrichment analysis_
Four gene sets were used for pathway analysis. For SARS-CoV-2 infection versus control, we defined a set of upregulated genes (FDR < 10%, 100% increase in expression, 178 genes) and downregulated genes (FDR < 10%, 50% reduction in expression, 23 genes). For SARS-CoV-2 dosage sensitive genes, we used only the samples with SARS-CoV-2 infection and correlated the expression level of the SARS-CoV-2 E and N genes (as calculated by qPCR) with gene expression, and identified correlated (R > 0.8) and anti-correlated (R < -0.8) gene sets. We took the intersection of genes for the E and N genes as the final set of dose-responsive genes. Gene Ontology pathway term enrichment was performed using TopGO (50). Cell-states were identified using data from the supplementary tables in Refs (18, 19). The LPS-stimulated activated microglia set was defined from Table S2 in Drager et al., using genes with fold change > 2 and FDR < 5%. The additional cell states were defined using the cluster-specific genes from the CROP-seq experiment delineated in Table S5 from that study. The IL-alpha/TNF/C1q stimulated reactive astrocyte set was defined from Table S1 in Leng et al., with fold change > 2 and FDR < 5%. The reactive astrocyte states were defined using the cluster-specific genes from CROP-seq in Table S8 from that study. Cell state enrichments were computed with a two-sided Fisherâ€™s exact test using all genes expressed in the bulk RNA-seq data as background. The heatmap plot shows only the odds-ratio values for enrichment with FDR < 1% for display purposes.
